binary_search_tree/tree.py

```diff
@@ -9,50 +9,124 @@ def __init__(self, key, val = None):
         self.right = None
         
 
-
 class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(logn)
+    # Space Complexity: O(1)
     def add(self, key, value = None):
-        pass
+        new_node = TreeNode(key, value)
+        if not self.root:
+            self.root = new_node
+            return None
+
+        curr = self.root
 
-    # Time Complexity: 
-    # Space Complexity: 
+        while True:
+            if curr.key > key and curr.left is None:
+                curr.left = new_node
+                return None
+            elif curr.key <= key and curr.right is None:
+                curr.right = new_node
+                return None
+            elif curr.key > key:
+                curr = curr.left
+            
+            else:
+                curr = curr.right
+
+    # Time Complexity: O(logn)
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        curr = self.root
+
+        while curr:
+            if curr.key == key:
+                return curr.value
+            elif curr.key > key:
+                curr = curr.left
+            else:
+                curr = curr.right
+        return None 
+        
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n) with n being the number of nodes
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        if not self.root:
+            return []
+
+        left_tree = Tree()
+        left_tree.root = self.root.left
+        right_tree = Tree()
+        right_tree.root = self.root.right
+
+        return left_tree.inorder() + [{"key": self.root.key, "value": self.root.value}] + right_tree.inorder()
+        
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n) with n being the number of nodes
+    # Space Complexity: O(n)   
     def preorder(self):
-        pass
+        if not self.root:
+            return []
 
-    # Time Complexity: 
-    # Space Complexity:     
+        left_tree = Tree()
+        left_tree.root = self.root.left
+        right_tree = Tree()
+        right_tree.root = self.root.right
+
+        return [{"key": self.root.key, "value": self.root.value}] + left_tree.preorder() + right_tree.preorder()
+
+    # Time Complexity: O(n) with n being the number of nodes
+    # Space Complexity: O(n)    
     def postorder(self):
-        pass
+        if not self.root:
+            return []
+
+        left_tree = Tree()
+        left_tree.root = self.root.left
+        right_tree = Tree()
+        right_tree.root = self.root.right
+
+        return left_tree.postorder() + right_tree.postorder() + [{"key": self.root.key, "value": self.root.value}]
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n) with n being the number of nodes
+    # Space Complexity: O(n)      
     def height(self):
```

Comment:
👍  However the space complexity is O(log n) for a balanced tree or O(h) where h is the height of the tree.